src/AppChooser/Dialog.vala

```diff
@@ -127,32 +128,55 @@ public class AppChooser.Dialog : Hdy.Window {
         };
         open_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
 
+        var top_grid = new Gtk.Grid () {
+            orientation = Gtk.Orientation.VERTICAL,
+            column_spacing = 12,
+            row_spacing = 6,
+            margin_start = 12,
+            margin_end = 12,
+            margin_top = 12
+        };
+
+        top_grid.attach (overlay, 0, 0, 1, 2);
+        top_grid.attach (primary_label, 1, 0);
+        top_grid.attach (secondary_label, 1, 1);
+
+        var window_handle_top = new Hdy.WindowHandle ();
+        window_handle_top.add (top_grid);
+
+        var content_grid = new Gtk.Grid () {
+            orientation = Gtk.Orientation.VERTICAL,
+            column_spacing = 12,
+            row_spacing = 6,
+            margin_start = 12,
+            margin_end = 12
+        };
+
+        content_grid.attach (frame, 0, 0);
+
         var button_box = new Gtk.ButtonBox (Gtk.Orientation.HORIZONTAL) {
             layout_style = Gtk.ButtonBoxStyle.END,
-            margin_top = 12,
+            margin = 12,
             spacing = 6
         };
 
         button_box.add (cancel);
         button_box.add (open_button);
 
+        var window_handle_bottom = new Hdy.WindowHandle ();
+        window_handle_bottom.add (button_box);
+
         var grid = new Gtk.Grid () {
             orientation = Gtk.Orientation.VERTICAL,
             column_spacing = 12,
-            row_spacing = 6,
-            margin = 12
```

Comment:
why remove the margin here? the logic here is that this set the external margins, so that the header and button box only need to deal with the internal margins (margin_botton for the header and margin_top for the button box).

Reply:
Leaving this 12px means the 4 edges of the dialog won't receive the key press events like the secondary click for the context menu and dragging for window movement. So I moved all margins into each grid. I'm not sure if these changes I made is correct though.